Use containerd to handle k8s.gcr.io deprecation (#40528)
kubernetes/website: content/en/blog/_posts/2023-04-07handle-registry-change-wwith-containerd.md

---
layout: blog
title: "Use containerd to handle k8s.gcr.io deprecation"
date: 2023-04-07
slug: handle-registry-deprecation-with-containerd
---

The Kubernetes community recently made the last step in another major change. Until fall of 2022, `k8s.gcr.io`
container registry was the official place to find many Kubernetes community-managed containers images;
for example, kube-proxy, the cluster autoscaler, the metrics server, or the cluster-proportional autoscaler.
The GCR in “gcr.io” is Google Cloud Registry. The Kubernetes project
[switched](/blog/2022/11/28/registry-k8s-io-faster-cheaper-ga/) in November 2022
to a new registry with a different implementation.

As a result, starting March 20th this year, traffic from the old `k8s.gcr.io` registry is being redirected to `registry.k8s.io`. The deprecated `k8s.gcr.io` registry will remain functioning for some time but may one day be shut down entirely.

> {{< figure src="2023-04-07handle-registry-change-with-containerd/00-registry-deprecation-announcement.png" alt="K8s.gcr.io registry deprecation notice" class="diagram-large" caption="K8s.gcr.io registry deprecation notice" >}}

Reviewer: ```suggestion
{{< figure src="2023-04-07handle-registry-change-with-containerd/00-registry-deprecation-announcement.png" alt="Legacy k8s.gcr.io container image registry is being redirected to registry.k8s.io" class="diagram-large" caption="k8s.gcr.io registry redirection notice" >}}
```

Once the old registry goes away, **Kubernetes will not be able to create new Pods** (unless image is cached) if the container uses an image hosted on `k8s.gcr.io`.

## What do you need to change?

Cluster owners and development teams have to ensure they are not using any images stored in the old registry. The change is fairly simple. It's pretty simple, really.

You need to change your manifests to use the `registry.k8s.io` container registry instead of `k8s.gcr.io`.

{{< figure src="2023-04-07handle-registry-change-with-containerd/01-registry-change.png" alt="Changing image registry from k8s.gcr.io to registry.k8s.io" class="diagram-large" caption="Changing image registry from k8s.gcr.io to registry.k8s.io" >}}

You can find out which Pods use the old registry using `kubectl`:

```bash
kubectl get pods --all-namespaces -o jsonpath="{.items[*].spec.containers[*].image}" |\0;31;37M0;31;38m
tr -s '[[:space:]]' '\n' |\
sort |\
uniq -c | grep -i gcr.io
```

Here are the Pods in my test cluster that use the old registry:

> {{< figure src="2023-04-07handle-registry-change-with-containerd/02-images-from-gcr.png" alt="Containers using k8s.gcr.io registry" class="diagram-large" caption="Containers using k8s.gcr.io hosted images" >}}

Reviewer: {{< figure src="/blog/_posts/2023-04-07handle-registry-change-with-containerd/06-cotainer works.png/02-images-from-gcr.png" alt="Containers using k8s.gcr.io registry" class="diagram-large" caption="Containers using k8s.gcr.io hosted images" >}}

SystemdCgroup = true

[plugins."io.containerd.grpc.v1.cri".cni]
bin_dir = "/opt/cni/bin"
conf_dir = "/etc/cni/net.d"

[plugins."io.containerd.grpc.v1.cri".registry]
config_path = "/etc/containerd/certs.d"
```

Next, create a directory called `k8s.gcr.io` and a `hosts.toml` file inside it:

```bash
mkdir -p /etc/containerd/certs.d/k8s.gcr.io

cat << EOF > /etc/containerd/certs.d/k8s.gcr.io/hosts.toml
server = "https://k8s.gcr.io"

[host."https://registry.k8s.io"]
capabilities = ["pull", "resolve"]
EOF
```

Image pull requests for `k8s.gcr.io` will now be sent to `registry.k8s.io`.

Restart containerd and kubelet for the change to take effect.

```bash
# This assumes that you use Linux nodes and use systemd for init
systemctl restart containerd kubelet
```

Let’s validate that images are indeed getting pulled from the new registry. I added an entry to my `/etc/hosts` file to break `k8s.gcr.io`:

{{< figure src="2023-04-07handle-registry-change-with-containerd/03-hosts-file-change.png" alt="Break k8s.gcr.io" class="diagram-large" caption="Stop traffic to k8s.gcr.io in the node" >}}

Containerd can no longer pull an image from `k8s.gcr.io`:

> {{< figure src="2023-04-07handle-registry-change-with-containerd/04-registry-broken.png" alt="Conatinerd cannot pull images from k8s.gcr.io" class="diagram-large" caption="Connections k8s.gcr.io blocked at node level" >}}

Reviewer: ```suggestion
{{< figure src="2023-04-07handle-registry-change-with-containerd/04-registry-broken.png" alt="Screenshot showing that containerd cannot pull images from k8s.gcr.io" class="diagram-large" caption="Connections k8s.gcr.io blocked at node level" >}}
```

> I can tell `ctr` to use the mirror by specifying the `—hosts-dir` parameter:

Reviewer: ```suggestion
I can tell `ctr` to use the mirror by specifying the `--hosts-dir` parameter:
```
